app/cluster/state/etcd.go

```diff
@@ -0,0 +1,289 @@
+	manager.once.Do(func() {
+		cli, err := clientv3.New(clientv3.Config{
+			Endpoints:            manager.Config.Endpoints,
+			DialTimeout:          manager.Config.dialTimeout,
+			DialKeepAliveTime:    manager.Config.dialKeepAliveTime,
+			DialKeepAliveTimeout: manager.Config.dialKeepAliveTimeout,
+			DialOptions: []grpc.DialOption{
+				grpc.WithBlock(), // block until the underlying connection is up
+			},
+		})
+		if err != nil {
+			manager.initErr = err
+			return
+		}
+		manager.Client = cli
+	})
+
+	return manager.initErr
+}
+
+// Ping ensures the connection to etcd is alive
+func (manager *ETCDManager) Ping() error {
+	if err := manager.init(); err != nil {
+		return err
+	}
+
+	_, err := manager.Client.Cluster.MemberList(context.Background())
+	if err != nil {
+		return fmt.Errorf("ping: get cluster member list: %w", err)
+	}
+
+	return nil
+}
+
+func (manager *ETCDManager) prepareMode(raw []byte) servermode.ChangeEvent {
+	var req modeRequestValue
+	err := json.Unmarshal(raw, &req)
+	if err != nil {
+		return servermode.ChangeEventError(fmt.Errorf("unmarshal mode request: %w", err))
+	}
+
+	mode := req.Mode
+	if !mode.Valid() {
+		return servermode.ChangeEventError(fmt.Errorf("invalid mode: %s", mode))
+	}
+
+	return servermode.NewChangeEvent(
+		mode,
+		func() error {
+			ackValue, err := json.MarshalToString(modeAckValue{
+				Status: mode,
+			})
+			if err != nil {
+				return fmt.Errorf("marshal ack value: %w", err)
+			}
+			_, err = manager.Client.Put(context.Background(), req.AckKey, ackValue)
+			if err != nil {
+				return fmt.Errorf("put value to ack key %q: %w", req.AckKey, err)
+			}
+			return err
+		})
+}
+
+func errChModeRequest(err error) <-chan servermode.ChangeEvent {
+	ch := make(chan servermode.ChangeEvent, 1)
+	ch <- servermode.ChangeEventError(err)
+	close(ch)
+	return ch
+}
+
+func (manager *ETCDManager) ServerMode(ctx context.Context) <-chan servermode.ChangeEvent {
+	if err := manager.init(); err != nil {
+		return errChModeRequest(err)
+	}
+
+	modeRequestKey := fmt.Sprintf(modeRequestKeyPattern, manager.Config.Namespace, manager.Config.ServerIndex)
+
+	resultChan := make(chan servermode.ChangeEvent, 1)
+	resp, err := manager.Client.Get(ctx, modeRequestKey)
+	if err != nil {
+		return errChModeRequest(err)
+	}
+
+	if len(resp.Kvs) == 0 {
+		return errChModeRequest(errors.New("no workspace found"))
+	}
+
+	resultChan <- manager.prepareMode(resp.Kvs[0].Value)
+
+	etcdWatchChan := manager.Client.Watch(ctx, modeRequestKey, clientv3.WithRev(resp.Header.Revision+1))
+	go func() {
+		for watchResp := range etcdWatchChan {
+			if watchResp.Err() != nil {
+				resultChan <- servermode.ChangeEventError(watchResp.Err())
+				continue
+			}
+
+			for _, event := range watchResp.Events {
+				switch event.Type {
+				case mvccpb.PUT:
+					resultChan <- manager.prepareMode(event.Kv.Value)
+				default:
+					resultChan <- servermode.ChangeEventError(fmt.Errorf("unknown event type %q", event.Type))
```

Comment:
Do we plan to handle different errors from etcd layer differently on the dynamic cluster manager?

Reply:
Some errors like this should be pretty harmless where as the watch error might need some handling

Reply:
Yeah, I was thinking about this as well. 

Some errors like the parsing could probably be handled. 

We need to ensure that scheduler and other components will be ok with server ignoring some values.

For simplicity I kept it like this, until we revisit our error handling in a higher level.

Reply:
Contract should ensure that Scheduler does not return any other values apart from PUT value.So should be safe to ignore I guess